design/ironic-standalone-networking-service.md

```diff
@@ -0,0 +1,734 @@
+   updated after inspection.
+
+3. **BareMetalNetworkAttachment Immutability**: To prevent accidental
+   misconfiguration, users cannot modify or delete a `BareMetalNetworkAttachment`
+   while it is referenced by any `BareMetalHost`. The controller will reject
+   such operations.
+
+4. **Switch Configuration Changes**: When `BareMetalSwitch` resources are
+   modified, the entire switch configuration Secret is regenerated and the
+   Ironic Networking service is restarted. This may cause brief service
+   interruptions.
+
+5. **Credential Management**: Switch credentials are stored in Kubernetes
+   Secrets and referenced by `BareMetalSwitch` resources. The Metal³ controller
+   will read these Secrets to populate the switch configuration file.
+
+6. **Container Deployment**: The Ironic Networking service will run as a
+   separate container in the metal3 pod, alongside the existing metal3-ironic
+   container. It will share the `/etc/ironic` volume for configuration files
+   and the htpasswd file.
+
+7. **Network Driver Selection**: The Baremetal Operator will automatically set
+   the Ironic node's network interface driver to `standalone-experimental`
+   when:
+   - `enableIronicNetworking` is true in the Provision resource, AND
+   - The `BareMetalHost` has `networkInterfaces` defined
+
+8. **Service Network Scope**: The service network configuration applies to a
+   subset of the Ironic provider networks. Specifically, it configures only
+   the idle and inspection networks. All other provider networks (provisioning,
+   cleaning, rescuing, servicing) will default to the tenant VLAN configured on
+   each port. This means only two VLANs are needed: the service VLAN for
+   Ironic operations, and the tenant VLAN(s) configured per-port.
+
+9. **Generic Switch Device Types**: The `deviceType` field in `BareMetalSwitch`
+   must be one of the device types supported by networking-generic-switch.
+   Valid values include, but are not limited to, `cisco_ios`, `dell_force10`,
+   `juniper_junos`.  The full list can be found in the setup.cfg file of the
+   networking-generic-switch project.
+
+### Risks and Mitigations
+
+**Risk**: Metal³'s instantiation of Ironic currently does not include
+persistent storage for the Ironic database.  If the Ironic Pod is restarted
+then all node and port information is lost, including the LLDP information
+and switch port configuration derived from the BMH configuration.
+
+**Mitigation**: The BMH reconciler can be modified to restore the last known
+set of Port objects including their `port.extra.switchport` and
+`port.local_link_connection` attributes.
+
+**Risk**: Switch configuration errors could render hosts unreachable or prevent
+provisioning.
+
+**Mitigation**: Implement thorough validation of `BareMetalSwitch` and
+`BareMetalNetworkAttachment` resources. Add status conditions to report
+configuration errors. Ensure service network is properly configured before
+attempting to provision hosts.
```

Comment:
We need to make sure that if cleaning is disabled, we don't try to access switches or validating anything on BMH deletion. So that even broken hosts can be cleanly deleted.

Reply:
I _think_ that's ok... with my latest revision in Ironic setting the port's VLAN back to the "idle" VLAN should happen naturally without any extra steps so when transitioning into a state where delete would be accepted the switch port should already be ok.